replica/table.cc

```diff
@@ -114,17 +114,20 @@ table::make_sstable_reader(schema_ptr s,
 }
 
 sstables::sstable_set compaction_group::make_compound_sstable_set() {
-    return sstables::make_compound_sstable_set(_t.schema(), { _main_sstables, _maintenance_sstables });
+    return sstables::make_compound_sstable_set(_t.schema(), { &_main_sstables, &_maintenance_sstables });
 }
 
 sstables::sstable_set table::make_compound_sstable_set() {
     if (auto cg = single_compaction_group_if_available()) {
         return cg->make_compound_sstable_set();
     }
     // TODO: switch to a specialized set for groups which assumes disjointness across compound sets and incrementally read from them.
-    // FIXME: avoid recreation of compound_set for groups which had no change. usually, only one group will be changed at a time.
-    auto sstable_sets = boost::copy_range<std::vector<sstables::sstable_set>>(compaction_groups()
-        | boost::adaptors::transformed(std::mem_fn(&compaction_group::make_compound_sstable_set)));
+    std::vector<const sstables::sstable_set*> sstable_sets;
+    sstable_sets.reserve(2 * compaction_groups().size());
+    for (const auto& cg : compaction_groups()) {
+        sstable_sets.emplace_back(&cg->main_sstables());
+        sstable_sets.emplace_back(&cg->maintenance_sstables());
+    }
     return sstables::make_compound_sstable_set(schema(), std::move(sstable_sets));
```

Comment:
this is a good change. avoids one layer of indirection that comes with compound set. for tablet aware sstable set, we'll only probably need only a hint that sets belong to a single group. it's probably good to ask group for a vector of sets. so the knowledge of which sets are used is confined to compaction_group.

Reply:
The thing is that in this branch we update the sets in-place so even if e.g. some compaction_group maintenance sstable set is now empty, it might be populated later and we don't want to rebuild the table's compound sstable_set for that.